### ISSUE TITLE
has_one association auto create doesn't work when use_schemas = false

### ISSUE THREAD

**Issue by fernandomm** [influitive/apartment#441]
I'm having an issue with has_one association when use_schemas is set to false. I only tried it with Mysql, so I don't know if it happens with PostgreSQL.

Here are some sample models:

```
class Account < ActiveRecord::Base
  has_one :package

  before_create :auto_create_package

  def auto_create_package
    self.package = build_package({
      name: 'Test'
    })
  end
end

class Package < ActiveRecord::Base
  belongs_to :account
end

```

Both models are excluded in apartment configuration:

```
config.excluded_models = %w{ Account Package }
```

When i try to create an account, it will hang indefinitely in package creation:

```
> a = Account.new({name: 'test'}); a.save
   (0.3ms)  BEGIN
  SQL (1.6ms)  INSERT INTO `accounts` (`name`, `created_at`, `updated_at`) VALUES ('test', '2017-06-23 17:01:56', '2017-06-23 17:01:56')
   (0.3ms)  BEGIN
```

Or, depending on server configuration, it will raise the following error after some time:

```
Lock wait timeout exceeded; try restarting transaction: INSERT INTO `packages` (`name`, `account_id`, `c
reated_at`, `updated_at`) VALUES ('Test', 7, '2017-06-23 17:01:56', '2017-06-23 17:01:56')
```

If I enable use_schemas, it works fine:

```
> a = Account.new({name: 'test'}); a.save
   (0.2ms)  BEGIN
  SQL (0.4ms)  INSERT INTO `one_bug`.`accounts` (`name`, `created_at`, `updated_at`) VALUES ('test', '2017-06-23 17:03:04', '2017-06-23 17:03:04')
  SQL (0.4ms)  INSERT INTO `one_bug`.`packages` (`name`, `account_id`, `created_at`, `updated_at`) VALUES ('Test', 8, '2017-06-23 17:03:04', '2017-06-23 17:03:04')
   (0.4ms)  COMMIT
```

I can create a repository with a reproducible example if it helps. But the issue seems to be that with use_schemas = false, it creates a new transaction when package is created, instead of reusing the current transaction.

**Comment by mikecmpbll** [comment]
it can't reuse the current transaction because the models use different database connections. even though they're both excluded, a new connection to the default tenant is established separately for each model: https://github.com/influitive/apartment/blob/development/lib/apartment/adapters/abstract_adapter.rb#L125-L127

this is fixed in my 5.1 only fork that i'm currently working on. i will be publishing it to test by the end of the week, and it includes a lot of refactoring on the adapters.

**Comment by fernandomm** [comment]
I forgot to post the solution for this issue but here it is.

This issue was happening because Account and Package models were both excluded models. So, Apartment was creating a new connection for each excluded model, exactly where @mikecmpbll indicated.

Since packages.account_id column depended on the value of account_id which was created in another transaction ( the one initiated by Account connection ) and it was never finished, MySQL would hang or fail because the account_id didn't existed for Package model connection/transaction.

I was able to fix this by leaving config.excluded_models empty and making Account and Package models inherit from the same abstract class, with a establish_connection method call using my main database details.

Example:

```
class MainDatabase < ActiveRecord::Base
  self.abstract_class = true

  establish_connection Rails.configuration.database_configuration[Rails.env]
end

class Account < MainDatabase
  has_one :package

  before_create :auto_create_package

  def auto_create_package
    self.package = build_package({
      name: 'Test'
    })
  end
end

class Package < MainDatabase
  belongs_to :account
end
```

This fixed the issue because Account and Package model now share the same connection and transactions.
